state/indexer/block/kv/util.go

```diff
@@ -74,24 +75,67 @@ func parseValueFromEventKey(key []byte) (string, error) {
 		height                        int64
 	)
 
-	remaining, err := orderedcode.Parse(string(key), &compositeKey, &eventValue, &height, &typ)
+	_, err := orderedcode.Parse(string(key), &compositeKey, &eventValue, &height, &typ)
 	if err != nil {
 		return "", fmt.Errorf("failed to parse event key: %w", err)
 	}
 
+	return eventValue, nil
+}
+
+func parseHeightFromEventKey(key []byte) (int64, error) {
+	var (
+		compositeKey, typ, eventValue string
+		height                        int64
+	)
+
+	_, err := orderedcode.Parse(string(key), &compositeKey, &eventValue, &height, &typ)
+	if err != nil {
+		return -1, fmt.Errorf("failed to parse event key: %w", err)
+	}
+
+	return height, nil
+}
+
+func parseEventSeqFromEventKey(key []byte) (int64, error) {
+	var (
+		compositeKey, typ, eventValue string
+		height                        int64
+		eventSeq                      int64
+	)
+
+	remaining, err := orderedcode.Parse(string(key), &compositeKey, &eventValue, &height, &typ)
+	if err != nil {
+		return 0, fmt.Errorf("failed to parse event key: %w", err)
+	}
+
+	// This is done to support previous versions that did not have event sequence in their key
+
 	if len(remaining) != 0 {
-		return "", fmt.Errorf("unexpected remainder in key: %s", remaining)
+		remaining, err = orderedcode.Parse(remaining, &eventSeq)
+		if err != nil || len(remaining) != 0 {
```

Comment:
It's better to separate these conditions into two "if"s, because the error messages will be much clearer.
* If "err != nil" you will want to include the error in the message
* If remaining is not empty, you will want to the contents of remaining.
Besides, the error message can be clearer. Check lines 77-84 of the OLD version of this file (the version to the left of the diff)